### ISSUE TITLE
Allow SVG's to have 0 width and height on root element

### ISSUE THREAD

**Issue by jakepruitt** [mapnik/mapnik#3077]
According to the [SVG specification](http://www.w3.org/TR/SVG11/coords.html#ObjectBoundingBox), in the event of missing height or width on `<svg>` tags, `viewBox` is used by default:

> If either/both of the ‘width’ and ‘height’ of the rootmost ‘svg’ element are in percentage units (or omitted), the aspect ratio is calculated from the width and height values of the ‘viewBox’ specified for the current SVG document fragment. If the ‘viewBox’ is not correctly specified, or set to 'none', the intrinsic aspect ratio cannot be calculated and is considered unspecified.

This contradicts how we currently parse SVG dimensions, which defaults to 0 when no width or height is provided.

cc/ @artemp @springmeyer @flippmoke

**Comment by artemp** [comment]
@jakepruitt - great catch! Are you planning to implement "viewBox" parsing and handling or should I take of it ?

**Comment by jakepruitt** [comment]
@artemp I wanted to give at an initial sketch today if you don't mind. I was going to add it to the `parse_dimensions` [code](https://github.com/mapnik/mapnik/blob/master/src/svg/svg_parser.cpp#L464), falling back on viewBox parsing if width or height are in percentage units or omitted. I will probably need to create a new parser for the four-digit viewBox format. Does this sound like the right way to proceed?

**Comment by artemp** [comment]
@jakepruitt - sounds great, let me know if you need any help coding simple boost::spirit::qi based parser.

**Comment by jakepruitt** [comment]
After spending some time with the spec, I think the logic for this function should work this way:

```
if (width and height are in px units):
  set width and height to width and height
else if (width is in px units and height is percentage/undefined and ViewBox exists):
  calculate aspect ratio from ViewBox, calculate height = width/aspect ratio
else if (height is in px units and width is percentage/undefined and ViewBox exists):
  calculate aspect ratio from ViewBox, calculate width = height*aspect ratio
else if (width and height are percentages/undefined and ViewBox exists):
  set width = ViewBox width, height = ViewBox height
```

If you read this closely, you'll notice that the percentage values of width and height are never used in the calculation of the width or height. This is because of [the following sentence in the spec](http://www.w3.org/TR/SVG/coords.html#IntrinsicSizing), which still bothers me a bit:

> Specifically, percentage values do not provide an intrinsic width or height, and do not indicate a percentage of the containing block. Rather, once the viewport is established, they indicate the portion of the viewport that is actually covered by image data.

I'm not quite sure how to take this, but my gut feeling is that percentages do not play a role in determining intrinsic width or height, which would be set by the `width` and `height` properties in the standard case. I'm going to move forward with this logic unless you think otherwise @artemp.

Also, in the interest of keeping things simple, I'm not going to implement the features of scaling the coordinate system that take place when a valid width/height **and** a ViewBox property are set. The spec is very convoluted on this issue, so I will maintain the current behavior of ignoring the ViewBox in that case.

**Comment by tmcw** [comment]
@jakepruitt is this fixed in #3081?

**Comment by springmeyer** [comment]
@tmcw yes, fixed by #3081 so I'll close this ticket. For reference this landed in Mapnik v3.0.6 (https://github.com/mapnik/mapnik/blob/master/CHANGELOG.md#306)
